### PR TITLE
[Example] add ogc method

### PR DESCRIPTION
**Pull request by SinuoXu**
## Description

 An implementation of OGC method from the paper [From Cluster Assumption to Graph Convolution: Graph-based Semi-Supervised Learning Revisited](https://arxiv.org/abs/2309.13599). 

OGC is a simple but very powerful graph convolution method, with only one trainable layer. Our initial proposal of the model mainly contains two files: utils.py which defines some basic functions, and main.py which shows the training progress.

**Table 1: Node classification accuracy (%).**

|    Acc    | CiteSeer | Cora | PubMed |
| :-------: | :------: | :--: | :----: |
|    GCN (DGL)    |   71.0   | 81.0 |  79.0  |
| GRAND (DGL) | 74.0 | 83.7 | 78.8 |
| GATv2 (DGL)| 70.3 | 81.0 | 77.1 |
| OGC (DGL) |   77.4 | **86.9** |  **83.6**  |
| OGC (Reported) |   **77.5**   | **86.9** |  83.4  |

**Table 2: Running time (s).**

|   Time    | CiteSeer | Cora | PubMed |
| :-------: | :------: | :--: | :----: |
|    GCN (DGL)   |   7.8   | 4.7 |  9.7  |
| GRAND (DGL) | 248.7 | 99.6 | 237.8 |
| GATv2 (DGL) | 17.51 | 11.71 | 44.44|
| OGC (DGL) |   **2.18**   | **0.95** |  **1.87**  |

We use three DGL's built_in datasets (Citeseer, Cora and Pubmed) with their default train/val/test settings. As shown above, OGC has a substantial improvement, with 100-200X high efficiency compared to SOTA deep GNNs.

We are very happy to further improve the code for this great project, and will continue to contribute actively!

Author: [Sinuo Xu](https://github.com/SinuoXu) (undergraduate, @SJTU) advised by [Dr. Zheng Wang@SJTU](https://zhengwang100.github.io/).

## Checklist
Please feel free to remove inapplicable items for your PR.
- [√] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [√] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [√] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [√] All changes have test coverage
- [√] Code is well-documented
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: ogc.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
+import time
```
Comment (Rhett-Ying) [inline ogc.py:1]: remove file from root dir

File: examples/pytorch/ogc/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Optimized Graph Convolution (OGC)
+
+This DGL example implements the OGC method from the paper: [From Cluster Assumption to Graph Convolution: Graph-based Semi-Supervised Learning Revisited](https://arxiv.org/abs/2309.13599).
+With only one trainable layer, OGC is a very simple but powerful graph convolution method.
+
+
+## Example Implementor
```
Comment (frozenbugs) [inline examples/pytorch/ogc/README.md:7]: Can you mention in Readme about what dgl version has been tested with this example?

**Review by frozenbugs — APPROVED** [review]

File: examples/pytorch/ogc/ogc.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+from dgl.data import CiteseerGraphDataset, CoraGraphDataset, PubmedGraphDataset
+from utils import sparse_mx_to_torch_sparse_tensor, symmetric_normalize_adjacency, LinearNeuralNetwork
+
+
+# Training settings
+decline = 0.9            # the dcline rate
+eta_sup = 0.001          # the learning rate for supervised loss
+eta_W = 0.5              # the learning rate for updating W
+beta = 0.1               # in [0,1], the moving probability that a node moves to its neighbors
+max_similar_tol = 0.995  # the max_tol test set label prediction similarity between two iterations
```
Comment (frozenbugs) [inline examples/pytorch/ogc/ogc.py:18]: nit: max line width: 80

Comment (frozenbugs) [inline examples/pytorch/ogc/ogc.py:18]:
```suggestion
max_similar_tol = 0.995  # The max_tol test set label prediction similarity
                         # between two iterations.
```

File: examples/pytorch/ogc/ogc.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+import torch
+import torch.nn.functional as F
+
+from dgl import AddSelfLoop
+from dgl.data import CiteseerGraphDataset, CoraGraphDataset, PubmedGraphDataset
+from utils import sparse_mx_to_torch_sparse_tensor, symmetric_normalize_adjacency, LinearNeuralNetwork
+
+
+# Training settings
+decline = 0.9            # the dcline rate
```
Comment (frozenbugs) [inline examples/pytorch/ogc/ogc.py:14]: nit, typo.
start with uppercase and end with dot.

Comment (frozenbugs) [reply examples/pytorch/ogc/ogc.py:14]: same for others.

File: examples/pytorch/ogc/ogc.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+eta_sup = 0.001          # the learning rate for supervised loss
+eta_W = 0.5              # the learning rate for updating W
+beta = 0.1               # in [0,1], the moving probability that a node moves to its neighbors
+max_similar_tol = 0.995  # the max_tol test set label prediction similarity between two iterations
+max_patience = 2         # the tolreance for consecutively getting very similar test prediction
+
+
+def update_U(U, Y, predY, W):
+    global eta_sup
+    # ------ update the smoothness loss via LGC ------
```
Comment (frozenbugs) [inline examples/pytorch/ogc/ogc.py:24]: ditto. start with uppercase and end with dot.

File: examples/pytorch/ogc/ogc.py (hunk 3 above)
Comment (frozenbugs) [inline examples/pytorch/ogc/ogc.py:18]: avoiding using too many abbreviations makes it easier for newcomer to read.

**Comment by frozenbugs** [comment]
@dgl-bot

File: ogc.py (hunk 1 above)
Comment (SinuoXu) [reply ogc.py]: Thanks. I've removed the file from root dir.

File: examples/pytorch/ogc/README.md (hunk 2 above)
Comment (SinuoXu) [reply examples/pytorch/ogc/README.md:7]: Thanks. We use DGL 1.1.2, and we have clearified all the dependencies in the readme file.

File: examples/pytorch/ogc/ogc.py (hunk 3 above)
Comment (SinuoXu) [reply examples/pytorch/ogc/ogc.py]: Thanks. We've fixed these formatting issues.

File: examples/pytorch/ogc/ogc.py (hunk 4 above)
Comment (SinuoXu) [reply examples/pytorch/ogc/ogc.py]: Thanks. We've adjusted all our annotations in accordance with the style guide.

File: examples/pytorch/ogc/ogc.py (hunk 5 above)
Comment (SinuoXu) [reply examples/pytorch/ogc/ogc.py]: Thanks. We have adjusted all of them.

**Comment by frozenbugs** [comment]
@BarclayII can you double check the implementation?

**Comment by frozenbugs** [comment]
@dgl-bot

File: examples/pytorch/ogc/ogc.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    graph = data[0]
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    graph = graph.int().to(device)
+    features = graph.ndata["feat"]
+
+    adj = symmetric_normalize_adjacency(graph)
+    I_N = sp.eye(features.shape[0])
+    # Lazy random walk (also known as lazy graph convolution).
+    lazy_adj = (1 - beta) * I_N + beta * adj
+    lazy_adj = sparse_mx_to_torch_sparse_tensor(lazy_adj)
```
Comment (frozenbugs) [inline examples/pytorch/ogc/ogc.py:101]: lazy_adj is not passed explicitly to update_embeds, which may cause confusion.

File: examples/pytorch/ogc/utils.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        np.vstack((sparse_mx.row, sparse_mx.col)).astype(np.int64)
+    )
+    values = torch.from_numpy(sparse_mx.data)
+    shape = torch.Size(sparse_mx.shape)
+    return torch.sparse.FloatTensor(indices, values, shape)
+
+
+def symmetric_normalize_adjacency(graph):
+    """Symmetric normalize graph adjacency matrix."""
+    adj = graph.adjacency_matrix()
```
Comment (frozenbugs) [inline examples/pytorch/ogc/utils.py:22]: Did you try using [dgl.sparse](https://docs.dgl.ai/en/1.1.x/api/python/dgl.sparse_v0.html) directly? Which may save you a lot of code from converting sparse format, and simplify the code for readability.

File: examples/pytorch/ogc/utils.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    in_degs = graph.in_degrees().float()
+    in_norm = torch.pow(in_degs, -0.5).unsqueeze(-1)
+    degi = torch.diag(torch.squeeze(torch.t(in_norm)))
```
Comment (BarclayII) [inline examples/pytorch/ogc/utils.py:25]: You could probably just do `in_norm.T` instead of `torch.t`.

File: examples/pytorch/ogc/ogc.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    deriv_sup = 2 * torch.mm(
+        torch.sparse.mm(label_idx_mat, -labels + pred_labels), clf_weight
+    )
+    embeds = embeds - lr_sup * deriv_sup
+
+    lr_sup = lr_sup * decline
+    return embeds
+
+
+def OGC(linear_clf, embeds, graph, label_idx_mat):
```
Comment (BarclayII) [inline examples/pytorch/ogc/ogc.py:46]: This is the training function but the function name is `OGC` which looks like a model class in terms of PyTorch convention.  This is confusing.  Consider renaming it to `train`.

File: examples/pytorch/ogc/ogc.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+# Training settings.
+decline = 0.9  # the decline rate
+lr_sup = 0.001  # the learning rate for supervised loss
+lr_clf = 0.5  # the learning rate for the used linear classifier
+beta = 0.1  # the moving probability that a node moves to its neighbors
+max_sim_rate = 0.995  # the max label prediction similarity between iterations
+max_patience = 2  # the tolerance for consecutively similar test predictions
+
+
+def update_embeds(embeds, graph, label_idx_mat):
```
Comment (BarclayII) [inline examples/pytorch/ogc/ogc.py:27]: This is a part of the OGC model, right?  If so, you could put it in an `OGC` NN module.

File: examples/pytorch/ogc/utils.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+class LinearNeuralNetwork(nn.Module):
+    def __init__(self, nfeat, nclass, bias=True):
+        super(LinearNeuralNetwork, self).__init__()
+        self.W = nn.Linear(nfeat, nclass, bias=bias)
+
+    def forward(self, x):
+        return self.W(x)
+
+    def test(self, embeds, g):
```
Comment (BarclayII) [inline examples/pytorch/ogc/utils.py:42]: This loss function computation belongs to the training loop in terms of PyTorch's convention as it is closely related to the downstream task.  It shouldn't be in the neural network module.

File: examples/pytorch/ogc/utils.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+                output[tv_mask], F.one_hot(labels).float()[tv_mask]
+            )
+            accs = []
+            for mask in [tv_mask, test_mask]:
+                accs.append(
+                    float((pred[mask] == labels[mask]).sum() / mask.sum())
+                )
+        return loss_tv.item(), accs[0], accs[1], pred
+
+    def update_weight(self, embeds, g, lr_clf):
```
Comment (BarclayII) [inline examples/pytorch/ogc/utils.py:60]: Same for optimization routine.

**Review by BarclayII — CHANGES_REQUESTED** [review]
The math looks simple and I don't have any problems.  My concerns are namely in code styles which is uncommon to PyTorch practices.  Normally in PyTorch we write training code like this:
```
class OGC(nn.Module):
    def __init__(self, ...):
        # model parameter definition...
    def forward(self, ...):
        # Given the input, how to compute the output (i.e. according to Eq 6 in your paper)
...
model = OGC(...)
def train(model, ...):
    optimizer = torch.optim.SGD(...)
    for _ in range(epochs):
        # compute loss...
        # call optimizer...
```

File: examples/pytorch/ogc/ogc.py (hunk 6 above)
Comment (SinuoXu) [reply examples/pytorch/ogc/ogc.py]: Thanks. As suggested, we have changed it.

File: examples/pytorch/ogc/utils.py (hunk 7 above)
Comment (SinuoXu) [reply examples/pytorch/ogc/utils.py]: Thank you for the excellent suggestion. By utilizing dgl.sparse, we have  effectively streamlined our normalization-related code in the "utils.py", reducing it from 21 lines to just 7 lines, all within a single function.

File: examples/pytorch/ogc/utils.py (hunk 8 above)
Comment (SinuoXu) [reply examples/pytorch/ogc/utils.py]: Thanks. As suggested, we have changed it.
And we've completely overhauled our codebase, adopting a 'gnn-pipeline' style. The updated version comprises three files: 'train.py' serves as the main file, 'ogc.py' houses the employed model, and 'utils.py' encompasses all fundamental method components.
Furthermore, in this part, we have refined the normalization process, aligning it with the example code in 'dgl/../pytorch/gcn'.

File: examples/pytorch/ogc/ogc.py (hunk 9 above)
Comment (SinuoXu) [reply examples/pytorch/ogc/ogc.py]: Thanks. As suggested, we have renamed it to 'train(..)', in the main file (i.e., the '[train.py](https://train.py/)' file).

File: examples/pytorch/ogc/ogc.py (hunk 10 above)
Comment (SinuoXu) [reply examples/pytorch/ogc/ogc.py]: Thanks. As suggested, we have incorporated this function into our OGC model in the 'ogc.py' file.

File: examples/pytorch/ogc/utils.py (hunk 11 above)
Comment (SinuoXu) [reply examples/pytorch/ogc/utils.py]: Thanks. As suggested, we have relocated the 'test' function, renamed as 'model_test' for clarity, to the 'utils.py' file as an independent function.

File: examples/pytorch/ogc/utils.py (hunk 12 above)
Comment (SinuoXu) [reply examples/pytorch/ogc/utils.py]: Thanks. As suggested, we have moved the optimization routine to the 'train' section in the main file.

**Comment by frozenbugs** [comment]
@dgl-bot
